server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java

```diff
@@ -125,18 +125,18 @@ public boolean isValid() {
         return getValidationErrors().isEmpty();
     }
 
-    public Flags applyToState(boolean isSeen, boolean isAnswered, boolean isFlagged) {
+    public Flags applyToState(Flags currentFlags) {
         Flags newStateFlags = new Flags();
 
-        boolean shouldMessageBeFlagged = isFlagged().isPresent() && isFlagged().get() || (!isFlagged().isPresent() && isFlagged);
+        boolean shouldMessageBeFlagged = isFlagged().isPresent() && isFlagged().get() || (!isFlagged().isPresent() && currentFlags.contains(Flags.Flag.FLAGGED));
```

Comment:
Ok, I find these  boolean expressions messy. If you touch to this class and wish to improve it, here is my feeling : 
 - I don't think I am supposed to know by heart which operator have priority between **&&** and **||**... Brackets will clarify this.
 - This line is repeated times. Its basically duplication. Why not introduce a function that takes an optional, and a boolean (here  *currentFlags.contains(Flags.Flag.FLAGGED)* ) and avoid repeating this logic.
 - Extract  each part of **||** to give it a name and better express logic...

Because like this I find it hard to know what goes on and we have already had this kind of requirement for other complicated boolean exceptions...